src/compas/datastructures/volmesh/core/halfface.py

```diff
@@ -268,7 +268,10 @@ def data(self, data):
             for u in cell[c]:
                 for v in cell[c][u]:
                     f = cell[c][u][v]
-                    faces.append(halfface[f])
+                    if f in halfface:
+                        faces.append(halfface[f])
+                    else:
+                        faces.append(halfface[str(f)])
```

Comment:
the if statement doesn't make sense to me. either it is supposed to be a string by construction or it is not. `f` is not a key in `cell`. therefore it will still be an `int` and should always be converted to `str` before using it in `halfface`. alternatively we could refer to the already processed `self._halfface` and just keep it as `int`

Reply:
I agree this is not a very good solution there, I m thinking about some better way.... Not sure if I understand all of it. But do you think we can just force all of the keys to be str? because once it's encoded to json and parsed back, it becomes very difficult to determine a key used to be a str or something else. Currently to_data() function produce a lot of int keys. This create a lot of ambiguity on how to parse them back from a json string.

Reply:
Or if a dictionary only suppose to have integer keys, we should consider use list instead

Reply:
we should just force all of the dict keys to `int` in the setter, because only `int` identifiers are supported by the data structure. using lists is not an option because then data management is a nightmare...

there should not be an issue. the mesh works in exactly the same way and there has never been a problem.